icarusalg/Utilities/AssnsCrosser.h

```diff
@@ -0,0 +1,1749 @@
+ *   (event, hopTo<DataTypeB>{ "B" }, hopTo<DataTypeC>{ "C" });
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * or
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * using icarus::ns::util::startFrom, icarus::ns::util::hopTo;
+ * icarus::ns::util::AssnsCrosser const AtoC{ event
+ *   , startFrom<DataTypeA>{}
+ *   , hopTo<DataTypeB>{ "B" }
+ *   , hopTo<DataTypeC>{ "C" }
+ *   };
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * or
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * using icarus::ns::util::startFrom, icarus::ns::util::hopTo;
+ * auto const AtoC = makeAssnsCrosser(event
+ *   , startFrom<DataTypeA>{}
+ *   , hopTo<DataTypeB>{ "B" }
+ *   , hopTo<DataTypeC>{ "C" }
+ *   );
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * The latter describe more clearly the relation between the data types and
+ * their input tags.
+ * 
+ * If there are two sets of associations between `DataTypeA` and `DataTypeB`,
+ * `"B:1"` and `"B:2"`, the following initializations will work:
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * icarus::ns::util::AssnsCrosser<DataTypeA, DataTypeB, DataTypeC> const AtoC
+ *   { event, { "B:1", "B:2" }, { "C" } };
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * or
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * using icarus::ns::util::hopTo;
+ * auto const AtoC = icarus::ns::util::makeAssnsCrosser<DataTypeA>(
+ *   event,
+ *   hopTo<DataTypeB>{ "B:1", "B:2" }, hopTo<DataTypeC>{ "C" }
+ *   );
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * 
+ * If the associations are needed only for a certain subset of pointers, it is
+ * possible to specify them, in a way similar to `art::FindManyP`:
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * using icarus::ns::util::startFrom, icarus::ns::util::hopTo;
+ * auto const AtoC = icarus::ns::util::makeAssnsCrosser<DataTypeA>(
+ *   , startFrom{ ptrA1, ptrA2 }
+ *   , hopTo<DataTypeB>{ "B" }
+ *   , hopTo<DataTypeC>{ "C" }
+ *   );
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * or
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * std::vector const selectedAptr { ptrA1, ptrA2 };
+ * using icarus::ns::util::startFrom, icarus::ns::util::hopTo;
+ * auto const AtoC = icarus::ns::util::makeAssnsCrosser<DataTypeA>(
+ *   , selectedAptr
+ *   , hopTo<DataTypeB>{ "B" }
+ *   , hopTo<DataTypeC>{ "C" }
+ *   );
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * Note that only a collection of type `std::vector<art::Ptr<DataTypeA>>` is
+ * supported (for example, `art::PtrVector<DataTypeA>` would not be).
+ */
```

Comment:
Possibly in this list of examples it could comment on the use of `assPtr` and `assPtrs` -- it is used for example in another PR I saw of yours and so was able to see an example usage there, or I guess in the unit tests here. Definitely not critical requirement for merging the code, but just a suggestion for a slight improvement to the documentation.

Reply:
Done in a new commit.
I wonder if it would be also useful to expose two derived classes for one-to-one and one-to-many associations which could feature either of `assPtr()` and `assPtrs()`, and in addition a `operator[]` (or `operator()`?).
E.g.
```cpp
/**
 * @brief `AssnsCrosser` class specific for one-to-one associations.
 * @tparam Key the type of the data to associate to
 * @tparam OtherTypes intermediate types to reach the target type (the last one)
 * @see `AssnsCrosser`
 *
 * This class only supports one-to-one associations between key type objects and target type objects.
 * The intermediate steps are not constrained to be one-to-one, but the final outcome is required to be
 * a single target type pointer at most.
 * 
 * In addition to the standard `AssnsCrosser` interface, an indexing `operator[]` is provided,
 * with the same functionality as `assPtr()`.
 */
template <typename KeyType, typename... OtherTypes>
class icarus::ns::util::AssnsCrosserOne: public icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...> {
  using Base_t = icarus::ns::util::AssnsCrosser<KeyType, OtherTypes...>;
  
  // not supported for one-to-one associations
  typename Base_t::TargetPtrs_t const& assPtrs(typename Base_t::KeyPtr_t const&) const = delete;

    public:
  using Base_t::Base_t; // use all inherited constructors

  /// Returns the pointer associated to `keyPtr` (alias of `assPtr()`), null if none.
  typename Base_t::TargetPtr_t const& operator[] (typename Base_t::KeyPtr_t const& keyPtr) const
    { return Base_t::assPtr(keyPtr); }
  
}; // icarus::ns::util::AssnsCrosserOne
```